Source/Core/VideoCommon/VertexShaderManager.cpp

```diff
@@ -327,10 +327,10 @@ void VertexShaderManager::SetConstants()
     const float* norm =
         &xfmem.normalMatrices[3 * (g_main_cp_state.matrix_index_a.PosNormalMtxIdx & 31)];
 
-    memcpy(constants.posnormalmatrix, pos, 3 * sizeof(float4));
-    memcpy(constants.posnormalmatrix[3], norm, 3 * sizeof(float));
-    memcpy(constants.posnormalmatrix[4], norm + 3, 3 * sizeof(float));
-    memcpy(constants.posnormalmatrix[5], norm + 6, 3 * sizeof(float));
+    memcpy(constants.posnormalmatrix.data(), pos, 3 * sizeof(float4));
```

Comment:
It seems suspicious to keep using memcpy with std::array, and it seems worse that the array is having multiple types copied into it.

Reply:
Yeah, the first memcpy is filling out 3 of the sub-arrays (I recall mentioning this looked a little odd to @degasus around two years or so ago). The following ones are just filling the rest individually, so it really is the same 'type', just a different way of filling it out. I can change them to all be uniform if you want.

Many of the memcpys end up being necessary for similar reasons (batch filling several arrays instead of on a by-array basis). I can unroll those and use std::copy instead if you want.

Reply:
Can't you just do the assignment with assignment operator?

Reply:
With a conversion for the float3-to-float4 case

Reply:
`xfmem.posMatrices` (what pos is a pointer into) is a `float[256]` and `xfmem.normalMatrices` (what norm is a pointer into) is a `float[96]`, so direct assignments can't be done here, unless I'm misinterpreting what you mean.

Reply:
something like:
```
constants.posnormalmatrix[0] = pos[0];
constants.posnormalmatrix[1] = pos[1];
constants.posnormalmatrix[2] = pos[2];
constants.posnormalmatrix[3] = float4(norm[0], norm[1], norm[2]);
constants.posnormalmatrix[4] = float4(norm[3], norm[4], norm[5]);
constants.posnormalmatrix[5] = float4(norm[6], norm[7], norm[8]);
```
...it /looks/ more clear, but maybe it's not worth it (maybe there's some better way - for example some way to reduce overall calls to memcpy).

Reply:
Yeah, I can change it to:

```cpp
constants.posnormalmatrix[0] = float4{{pos[0], pos[1], pos[2], pos[3]}};
constants.posnormalmatrix[1] = float4{{pos[4], pos[5], pos[6], pos[7]}};
constants.posnormalmatrix[2] = float4{{pos[8], pos[9], pos[10], pos[11]}};
constants.posnormalmatrix[3] = float4{{norm[0], norm[1], norm[2], 0.0f}};
constants.posnormalmatrix[4] = float4{{norm[3], norm[4], norm[5], 0.0f}};
constants.posnormalmatrix[5] = float4{{norm[6], norm[7], norm[8], 0.0f}};
```

Though that will zero out the 4th element of array in the `norm` cases, which wasn't done originally with the memcpys. I don't know if that would break anything or not though.

Reply:
I think you should just go with the existing changes, because breaking it into more clear assignments is likely to impact perf.